test/functional/Java16andUp/src/org/openj9/test/foreignMemoryAccess/TestCloseScope0.java

```diff
@@ -0,0 +1,164 @@
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+import org.testng.Assert;
+import static org.testng.Assert.*;
+import org.testng.annotations.Test;
+
+import java.lang.reflect.*;
+import java.lang.ref.Cleaner;
+
+import org.objectweb.asm.*;
+import static org.objectweb.asm.Opcodes.*;
+
+@Test(groups = { "level.sanity" })
+public class TestCloseScope0 {
+    public static byte[] dump() throws Exception {
+        ClassWriter cw = new ClassWriter(ClassWriter.COMPUTE_FRAMES);
+		MethodVisitor mv;
+		AnnotationVisitor av0;
+
+        cw.visit(V16, ACC_PUBLIC + ACC_SUPER, "jdk/internal/misc/RunInScoped", null, "java/lang/Object", null);
+        {
+			mv = cw.visitMethod(ACC_PUBLIC, "<init>", "()V", null, null);
+			mv.visitCode();
+			mv.visitVarInsn(ALOAD, 0);
+			mv.visitMethodInsn(INVOKESPECIAL, "java/lang/Object", "<init>", "()V");
+			mv.visitInsn(RETURN);
+            mv.visitMaxs(0, 0);
+			mv.visitEnd();
+		}
+        {
+			/**
+			 * @Scoped		// <-- package private annotation that isn't visible here.  This is why we use ASM to generate the class
+			 * public static void runInScoped(Runnable r, Scope scope) {
+			 *      r.run();
+             *      Reference.reachabilityFence(scope);
+			 * }
+			 */
+			mv = cw.visitMethod(ACC_PUBLIC + ACC_STATIC, "runInScoped", "(Ljava/lang/Runnable;Ljdk/internal/misc/ScopedMemoryAccess$Scope;)V", null, null);
+			{
+				av0 = mv.visitAnnotation("Ljdk/internal/misc/ScopedMemoryAccess$Scoped;", true);
+				av0.visitEnd();
+			}
+			mv.visitCode();
+            mv.visitVarInsn(ALOAD, 0);
+            mv.visitMethodInsn(INVOKEINTERFACE, "java/lang/Runnable", "run", "()V", true);
+            mv.visitVarInsn(ALOAD, 1);
+            mv.visitMethodInsn(INVOKESTATIC, "java/lang/ref/Reference", "reachabilityFence", "(Ljava/lang/Object;)V", false);
+			mv.visitInsn(RETURN);
+            mv.visitMaxs(0, 0);
+			mv.visitEnd();
+		}
+		cw.visitEnd();
+
+        return cw.toByteArray();
+    }
+
+    private static Throwable expected = null;
+
+    @Test(expectedExceptions=java.lang.IllegalStateException.class)
+    public static void closeScopeDuringAccess() throws Throwable {
+        /* Reflect setup */
+        Class c = Class.forName("jdk.internal.foreign.MemoryScope");
+        Method createShared = c.getDeclaredMethod("createShared", new Class[] {Object.class, Runnable.class, Cleaner.class});
+        createShared.setAccessible(true);
+        Method close = c.getDeclaredMethod("close");
+        close.setAccessible(true);
+        final Object scope = createShared.invoke(null, null, new Thread(), null);
```

Comment:
have you tested what happens with RI if scope isnt actually of type scope?

Reply:
RI:
```
Exception in thread "ScopeCloserThread"
java.lang.IllegalArgumentException: object is not an instance of declaring class
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:567)
        at Test.lambda$main$0(Test.java:89)
        at java.base/java.lang.Thread.run(Thread.java:831)
```
Our implementation:
```
Exception in thread "ScopeCloserThread" java.lang.IllegalArgumentException: object is not an instance of declaring class
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:567)
        at Test.lambda$main$0(Test.java:89)
        at Test$$Lambda$245/0x000000004491d120.run(Unknown Source)
        at java.base/java.lang.Thread.run(Thread.java:883)
```

Reply:
But when I run the RI with the original test it doesn't throw any exceptions but ours does.
RI:
```
root@purpurin1:~# jdk-16/bin/java --add-opens jdk.incubator.foreign/jdk.internal.foreign=ALL-UNNAMED --add-opens java.base/jdk.internal.misc=ALL-UNNAMED --add-opens java.base/java.lang=ALL-UNNAMED --add-modules jdk.incubator.foreign -cp openj9-openjdk-jdk16/openj9/test/functional/Java16andUp/module_bin/asm.jar openj9-openjdk-jdk16/Test.java
WARNING: Using incubator modules: jdk.incubator.foreign
warning: using incubating module(s): jdk.incubator.foreign
Note: openj9-openjdk-jdk16/Test.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: openj9-openjdk-jdk16/Test.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
1 warning
root@purpurin1:~# 
```
Ours:
```
root@purpurin1:~/openj9-openjdk-jdk16# build/linux-x86_64-server-release/images/jdk/bin/java --add-opens jdk.incubator.foreign/jdk.internal.foreign=ALL-UNNAMED --add-opens java.base/jdk.internal.misc=ALL-UNNAMED --add-opens java.base/java.lang=ALL-UNNAMED --add-modules jdk.incubator.foreign -cp openj9/test/functional/Java16andUp/module_bin/asm.jar Test.java
WARNING: Using incubator modules: jdk.incubator.foreign
warning: using incubating module(s): jdk.incubator.foreign
Note: Test.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Test.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
1 warning
Exception in thread "main" java.lang.IllegalStateException: Cannot close while another thread is accessing the segment
        at jdk.incubator.foreign/jdk.internal.foreign.MemoryScope$SharedScope.justClose(MemoryScope.java:282)
        at jdk.incubator.foreign/jdk.internal.foreign.MemoryScope.close(MemoryScope.java:97)
root@purpurin1:~/openj9-openjdk-jdk16# 
```

Reply:
Okay, its sounds like the RI is doing a type check on the scope object but not throwing anything if its invalid

Reply:
My second reply (RI not throwing IllegalStateException) was not testing with an invalid parameter. I ran the original test but RI was not throwing an exception when a scoped being accessed was closed. I think maybe their `closeScope0` function has not been implemented for 16?
```java
/**
 * Closes this scope, executing any cleanup action (where provided).
 * @throws IllegalStateException if this scope is already closed or if this is
 * a confined scope and this method is called outside of the owner thread.
 */
```

Reply:
> I think maybe their closeScope0 function has not been implemented for 16?

No its definitely implemented. This suggests that there may be an issue with our test

Reply:
I can't see anything wrong with the test. I put prints in the test and I am sure that the scope is closed while another thread is accessing the segment.